utilities/include/alps/numeric/tensors/data_storage.hpp

```diff
@@ -20,45 +52,154 @@ namespace alps {
        * @brief TensorBase class implements basic multi-dimensional operations.
        * All arithmetic operations are performed by Eigen library
        */
-      template<typename T>
+      template<typename T, typename Allocator = simple_allocator<T> >
       class data_storage {
       private:
         /// internal data storage
-        std::vector<T> data_;
+        T* data_;
+        /// size of buffer
+        size_t size_;
+        /// dataspace allocator
+        Allocator allocator_;
+        /// Check the data is locked in case of shared memory access
+        void check_lock() {
+          if(!allocator_.locked()) {
+            std::cerr<<"You are trying to access non-locked tensor with non-const method."<<std::endl;
+          }
+        }
```

Comment:
No way I will put throw there. If your object is non-const, it will always use non-const function. so you will always then have non-const access even if you are trying to read data.

Reply:
Ah okay, didn't think of that!